File: LeetCode/LeetCode 43. 字符串相乘 (高精乘).py
class Solution:
    def str_to_int(self, s: str) -> int:
        return ord(s) - ord('0')
    
    def multiply(self, num1: str, num2: str) -> str:
        a = num1[::-1]
        b = num2[::-1]
        ans = 0

        for i, x in enumerate(a):
            cur = 0
            for j, y in enumerate(b):
                cur += self.str_to_int(x) * self.str_to_int(y) * 10**j
            ans += cur * 10**i
        
        return str(ans)

File: LeetCode/LeetCode 43. 字符串相乘 (高精乘).cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        vector <int> ss(200, 0);
        vector <int> pp(200, 0);
        int n = num1.size(), m = num2.size();
        for(int i = n - 1; i >= 0; i--) {
            ss[n - i] = num1[i] - '0';
        }
        for(int i = m - 1; i >= 0; i--) {
            pp[m - i] = num2[i] - '0';
        }
        vector <int> ans(200, 0);
        for(int i = 1; i <= n; i++) {
            int x = 0;
            for(int j = 1; j <= m; j++) {
                ans[i + j - 1] += ss[i] * pp[j] + x;
                x = ans[i + j - 1] / 10;
                ans[i + j - 1] %= 10;
            }
            ans[i + m] = x;
        }
        int len = n + m;
        while(ans[len] == 0 && len > 1) len--;
        string now;
        for(int i = len; i > 0; i--) {
            now += ans[i] + '0';
        } 
        return now;
    }
};